pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/procfs"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/util/math"
+)
+
+const (
+	// Interval for updating resource (CPU/memory) utilization
+	resourceUtilizationUpdateInterval = time.Second
+
+	memPathV2 = "/sys/fs/cgroup/memory.current"
+	cpuPathV2 = "/sys/fs/cgroup/cpu.stat"
+	memPathV1 = "/sys/fs/cgroup/memory/memory.stat"
+	cpuPathV1 = "/sys/fs/cgroup/cpu/cpuacct.usage"
+)
+
+type utilizationScanner interface {
+	fmt.Stringer
+
+	// Scan returns CPU time in seconds and memory utilization in bytes, or an error.
+	Scan() (float64, uint64, error)
+}
+
+type cgroupV2Scanner struct {
+}
+
+func (s cgroupV2Scanner) String() string {
+	return "cgroup v2"
+}
+
+func (s cgroupV2Scanner) Scan() (float64, uint64, error) {
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
```

Comment:
What memory is this? We should be consistent between the different implementations, and I would suggest to use RSS for now.

According to https://facebookmicrosites.github.io/cgroup2/docs/memory-controller.html it includes the cache, which I would suggest to take in account. My take about cache is: as far as cache is not reclaimed, the process is not under memory pressure. Especially in the ingester, we may have the RSS + cache very high because of the cache, but that's OK, the kernel will use as much memory as it can for the cache. What we want to spot here is when the process is under memory pressure, which is when cache gets purged to reclaim memory, so just looking at RSS should give a better signal.

Reply:
As we discussed elsewhere, I've removed the cgroup logic.